service/history/historyEngine.go

```diff
@@ -1569,6 +1571,20 @@ func (e *historyEngineImpl) RespondActivityTaskFailed(
 				return nil, consts.ErrActivityTaskNotFound
 			}
 
+			cancelRequested = ai.CancelRequested
+
+			e.logger.Debug("RespondActivityTaskFailed", tag.WorkflowScheduleID(scheduleID), tag.ActivityInfo(ai), tag.NewBoolTag("cancelRequested", cancelRequested), tag.NewBoolTag("hasHeartBeatDetails", request.GetLastHeartbeatDetails() != nil))
+
+			if request.GetLastHeartbeatDetails() != nil && !request.GetLastHeartbeatDetails().Equal(ai.GetLastHeartbeatDetails()) {
+				// Save heartbeat details as progress
+				mutableState.UpdateActivityProgress(ai, &workflowservice.RecordActivityTaskHeartbeatRequest{
```

Comment:
I think with the call it's possible we will update and delete one activity in one transaction. and also generate a sync activity replication task for the deleted activity. Cost is quite small so I don't think it's high priority , we can probably create an issue to track it though.

Reply:
https://github.com/temporalio/temporal/pull/2572